### ISSUE TITLE
PC-98: Teitoku no Ketsudan has issues with text placement

### ISSUE THREAD

**Issue by cuba200611** [joncampbell123/dosbox-x#668]
So to test out DOSBox-X's PC-98 emulation, I started up Teitoku no Ketsudan - useful since it can be set to play by itself - and I have found some issues concerning text placement that is not found in Neko Project II.

For comparison:

DOSBox-X
![pc-98test2](https://user-images.githubusercontent.com/23426187/39015554-b8031386-43eb-11e8-9092-d69db6662527.PNG)
![pc-98test3](https://user-images.githubusercontent.com/23426187/39015560-bc195908-43eb-11e8-90e1-4ccd76eb4d52.PNG)

Neko Project II
![pc-98test4](https://user-images.githubusercontent.com/23426187/39015566-c226116a-43eb-11e8-8479-386b95544934.PNG)
![pc-98test5](https://user-images.githubusercontent.com/23426187/39015571-c4403e80-43eb-11e8-8eae-ade72d08654a.PNG)

**Comment by cuba200611** [comment]
The game comes in two disks, the first one which is bootable - but it uses disk-based copy protection...

**Comment by cuba200611** [comment]
Although I did find a way to extract the game files using NPII and some other tools...

I mounted the folder with the files to A:, and when I try to run it, this error pops up:
![pc-98test6](https://user-images.githubusercontent.com/23426187/39026171-8765f77e-4419-11e8-8077-057ed320b537.PNG)

So in order to make the game run, I have to mount that same folder to B: too.
Then it will run, although it still has issues with text formatting:
![pc-98test7](https://user-images.githubusercontent.com/23426187/39026224-e1468f10-4419-11e8-8b30-f15dbfd3378e.PNG)

**Comment by joncampbell123** [comment]
Looks like a problem with sending ANSI escapes.

Most games will send ANSI codes to clear the text screen and remove the function key row at the bottom, which apparently didn't happen in the last screenshot.

**Comment by yksoft1** [comment]
I think TK1 is one of Koei's more text-mode oriented games and made extensive uses of PC-98's text-mode tricks. Even the pointer on the strategy map (not tactics map) is drawn with text-mode characters I think.

**Comment by joncampbell123** [comment]
Are these text mode tricks in hardware or text mode tricks with MS-DOS's console driver?

PC-98 as far as I can tell provides only basic mode setup and cursor functions in the BIOS, and MS-DOS provides the TTY input/output with ANSI escapes and codes. Is this program repositioning the cursor with ANSI commands or by directly modifying the bytes in the DOS kernel related to the console?

There's a collection of "undocumented" information about PC-98 that includes fixed memory locations in the DOS kernel that some games will read directly. Touhou Project for example depends on a specific byte value within the DOS kernel to determine the number of character rows on the text display. It won't render the black border and game score correctly without it.

**Comment by yksoft1** [comment]
Some Koei's older game use their own custom DOS for copy protection/custom features. Many of its older games won't run or won't save/load data if ran on regular DOS or DOSBox-X's DOS. They might implemented some special features right in their custom DOS or not.

**Comment by timofonic** [comment]
@joncampbell123 About floppy emulation...

> DOSBox-X at this point is the worst emulator to use if you're using disk images that use copy protection. Maybe someday.

Related issues: #660 #444 #186

What about explaining in **_a separate issue_** the limitations of the current floppy emulation and what can/need to be improved? Something like a TODO list, but a bit more elaborated.

PS: I see this issue tracker has TOO MANY issues (252 at this moment), attention span isn't infinite and this is a bit "spammy". Maybe most of the issues could be joined as lists, because many of them are related to a specific architecture or a kind of issue. This way, I think they would be a lot more visible to everyone.

**Comment by joncampbell123** [comment]
@timofonic True, I need to stop treating the issue tracker like a TODO and FIX list.

**Comment by yksoft1** [comment]
There are still text placement problems in Teitoku no Ketsudan after the cursor position memory location fix. I booted the game using IMGMOUNT 0/1 BOOT -l way, so it should not be CON driver caused this.

![guest os_001](https://user-images.githubusercontent.com/37106774/41654797-1b75b7d4-747a-11e8-80ec-309de61b28ba.png)
![guest os_002](https://user-images.githubusercontent.com/37106774/41654798-1ba9f364-747a-11e8-81cc-f5241d2fa8df.png)
![guest os_003](https://user-images.githubusercontent.com/37106774/41654799-1be2d008-747a-11e8-8322-c9ff83eda174.png)
![guest os_004](https://user-images.githubusercontent.com/37106774/41654795-1b423102-747a-11e8-88d9-378072901528.png)

**Comment by yksoft1** [comment]
It runs even worse if I run it under Dosbox DOS. Seems a lot of ANSI commands used by TK1 is not emulated / not standard under NEC's DOS.
![renmain_000](https://user-images.githubusercontent.com/37106774/41660436-8f295bac-748b-11e8-9905-d2abfeaf15fe.png)
![renmain_001](https://user-images.githubusercontent.com/37106774/41660442-9401298e-748b-11e8-8d2e-9126281af7db.png)
![renmain_002](https://user-images.githubusercontent.com/37106774/41660449-9753376c-748b-11e8-99df-ed330e821799.png)

**Comment by yksoft1** [comment]
Seems that most misplacements are gone from TK1 after recent changes about CG and CON device.
@joncampbell123

**Comment by yksoft1** [comment]
> @yksoft1 Nice! How about color and attributes?
> 
> Also, I fixed FDC emulation use an alternate method that does not reprogram the timer. The latest commit still allows Ys II to run but fixes the issue with other games (in my tests, Angelus) running way too slow.

The text colors are also OK mostly. but there are still some differences compared with running on Neko II.
